apiserver/authwrapper/storage.go

```diff
@@ -0,0 +1,244 @@
+
+	// Storage returns the underlying storage
+	Storage() rest.Storage
+}
+
+type StorageScoper interface {
+	rest.Storage
+	rest.Scoper
+}
+
+// authorizedStorage is a wrapper around a rest.Storage
+// authorizing all requests and implementing the rest.Storage interface
+type authorizedStorage struct {
+	storage    rest.Storage
+	authorizer Authorizer
+}
+
+// authorizedStorageWithLister is a wrapper around a rest.StandardStorage that
+type authorizedStorageWithLister struct {
+	*authorizedStorage
+}
+
+// NewList implements rest.Lister/rest.StandardStorage
+func (s *authorizedStorageWithLister) NewList() runtime.Object {
+	return s.authorizedStorage.storage.(rest.Lister).NewList()
+}
+
+// ConvertToTable implements rest.Lister/rest.StandardStorage
+func (s *authorizedStorageWithLister) ConvertToTable(ctx context.Context, obj runtime.Object, tableOptions runtime.Object) (*metav1.Table, error) {
+	return s.authorizedStorage.storage.(rest.Lister).ConvertToTable(ctx, obj, tableOptions)
+}
+
+// NewAuthorizedStorage returns a new wrapper around the given storage
+// authorizing all requests based on rbacNamespace and implementing the rest.Storage or rest.StandardStorage interface.
+// It allows filtering list and watch results based on the user's RBAC permissions.
+// If the storage implements rest.StandardStorage, the returned storage will implement rest.StandardStorage.
+// If the storage implements rest.Storage, the returned storage will implement rest.Storage.
+// Only cluster-scoped resources currently are supported. Panics if the storage is namespace-scoped.
+func NewAuthorizedStorage(storage StorageScoper, rbacNamespace metav1.GroupVersionResource, auth authorizer.Authorizer) (Storage, error) {
+	if storage.NamespaceScoped() {
+		return nil, errors.New("namespace-scoped resources are not supported")
+	}
+	s := &authorizedStorage{
+		storage:    storage,
+		authorizer: NewAuthorizer(rbacNamespace, auth),
+	}
+	if _, ok := storage.(rest.Lister); ok {
+		return &authorizedStorageWithLister{s}, nil
+	}
+	return s, nil
+}
+
+// New implements rest.Storage
+func (s *authorizedStorage) New() runtime.Object {
+	return s.storage.New()
+}
+
+// Destroy implements rest.Storage
+func (s *authorizedStorage) Destroy() {
+	s.storage.Destroy()
+}
+
+// Storage returns the underlying storage
+func (s *authorizedStorage) Storage() rest.Storage {
+	return s.storage
+}
+
+// NamespaceScoped implements rest.Scoper
+// Only cluster-scoped resources are supported.
+func (s *authorizedStorage) NamespaceScoped() bool {
+	return false
+}
+
+func (s *authorizedStorage) Create(ctx context.Context, obj runtime.Object, createValidation rest.ValidateObjectFunc, opts *metav1.CreateOptions) (runtime.Object, error) {
+	err := s.authorizer.AuthorizeContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if stor, ok := s.storage.(rest.Creater); ok {
+		return stor.Create(ctx, obj, createValidation, opts)
+	}
+
+	return nil, newMethodNotSupported("create")
```

Comment:
Nitpicky nitpick: I'd prefer an early return for these methods, but it's fine as is